XLS-0065d - Single Asset Vault

maria-robobug opened the pull request:
Adds concept and reference documentation for [XLS-0065d Single Asset Vault](https://github.com/XRPLF/XRPL-Standards/blob/15180fc43162913a56981280e7cc68b839c4f038/XLS-0065d-single-asset-vault/README.md#321-vaultdeposit-transaction).

Note that the example JSON for transactions and ledger entries are not real. These will be updated when I get the chance to test the feature.

oeggert commented on docs/xls-65d-single-asset-vault/reference/vault.md:
```diff
@@ -0,0 +1,87 @@
+
+## Vault ID Format
+
+The key of the {% code-page-name /%} object is the [`SHA512-Half`](https://xrpl.org/docs/references/protocol/data-types/basic-data-types#hashes) of the following values, concatenated in order:
```
```suggestion
The ID of a {% code-page-name /%} entry is the [`SHA512-Half`](https://xrpl.org/docs/references/protocol/data-types/basic-data-types#hashes) of the following values, concatenated in order:
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-create.md:
```diff
@@ -0,0 +1,90 @@
+
+[[Source]](https://github.com/XRPLF/rippled/blob/9d619b9dc579c592f0560c1b40fd3c98d7587d23/src/xrpld/app/tx/detail/VaultCreate.cpp "Source")
+
+Creates a new `Vault` object, an `MPTokenIssuance` ledger entry for the vault’s shares, and an `AccountRoot` for the vault’s [pseudo-account](https://github.com/XRPLF/XRPL-Standards/discussions/191).
```
```suggestion
Creates a new `Vault` ledger entry, an `MPTokenIssuance` ledger entry for the vault’s shares, and an `AccountRoot` for the vault’s [pseudo-account](https://github.com/XRPLF/XRPL-Standards/discussions/191).
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-create.md:
```diff
@@ -0,0 +1,90 @@
+Only the Vault Owner can initiate this transaction.
+
+{% admonition type="info" name="Note" %}
+Currently, the same account that creates the vault must also create other protocols, though this may change in future.
```
```suggestion
Currently, the same account that creates the vault must also create other protocols, though this may change in the future.
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-create.md:
```diff
@@ -0,0 +1,90 @@
+{% admonition type="info" name="Note" %}
+Currently, the same account that creates the vault must also create other protocols, though this may change in future.
+{% /admonition %}
+
```
```suggestion
_(Requires the [Single Asset Vault amendment][] {% not-enabled /%})_
```
Just adding this in here to match the guidelines for easier doc migration later. This needs to be applied to all single asset vault pages. You'll also have to update the common-links.md file.

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-create.md:
```diff
@@ -0,0 +1,90 @@
+
+## WithdrawalPolicy
+
+A `WithdrawalPolicy` defines the strategy for processing withdrawal requests from a vault. This policy governs how liquidity is removed, and different strategies can be applied to different vaults, allowing each vault to have its own withdrawal policy. Currently, only one strategy is supported:
```
```suggestion
A `WithdrawalPolicy` defines the strategy for processing withdrawal requests from a vault. This policy governs how liquidity is removed. Currently, only one strategy is supported:
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-deposit.md:
```diff
@@ -0,0 +1,84 @@
+
+## Transfer Fees
+
+A Single Asset Vault does not apply the [Transfer Fee](https://xrpl.org/docs/concepts/tokens/transfer-fees) to {% code-page-name /%} transactions. Additionally, whenever a protocol moves assets from or to a vault, the Transfer Fee must not be charged.
```
```suggestion
A single asset vault does not apply the [transfer fee](https://xrpl.org/docs/concepts/tokens/transfer-fees) to {% code-page-name /%} transactions. Additionally, whenever a protocol moves assets from or to a vault, the transfer fee isn't charged.
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-set.md:
```diff
@@ -0,0 +1,63 @@
+
+[[Source]](https://github.com/XRPLF/rippled/blob/9d619b9dc579c592f0560c1b40fd3c98d7587d23/src/xrpld/app/tx/detail/VaultSet.cpp "Source")
+
+Updates any mutable field in an existing `Vault` ledger object.
+This transaction allows the Vault Owner to update certain mutable fields, including vault metadata and the maximum asset amount.
```
```suggestion
Modifies a single asset vault that you own..
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-withdraw.md:
```diff
@@ -0,0 +1,70 @@
+
+[[Source]](https://github.com/XRPLF/rippled/blob/9d619b9dc579c592f0560c1b40fd3c98d7587d23/src/xrpld/app/tx/detail/VaultWithdraw.cpp "Source")
+
+Withdraws assets from a vault or redeems vault shares for assets. The number of shares burned or assets received depends on the [exchange rate](../../concepts/single-asset-vault.md#exchange-algorithm), which adjusts based on the vault’s total assets and any [unrealized losses](../../concepts/single-asset-vault.md#paper-loss-unrealized-loss).
```
```suggestion
Redeem vault shares for assets. The number of shares burned or assets received depends on the [exchange rate](../../concepts/single-asset-vault.md#exchange-algorithm), which adjusts based on the vault’s total assets and any [unrealized losses](../../concepts/single-asset-vault.md#paper-loss-unrealized-loss).
```
It feels more precise to say you are always giving shares to receive assets.

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-withdraw.md:
```diff
@@ -0,0 +1,70 @@
+Withdraws assets from a vault or redeems vault shares for assets. The number of shares burned or assets received depends on the [exchange rate](../../concepts/single-asset-vault.md#exchange-algorithm), which adjusts based on the vault’s total assets and any [unrealized losses](../../concepts/single-asset-vault.md#paper-loss-unrealized-loss).
+
+{% admonition type="info" name="Note" %}
+The `VaultWithdraw` transaction does not respect the Permissioned Domain rules. In other words, any account that holds the shares of the vault can withdraw them. This is to avoid a situation where a depositor deposits assets to a private vault to then have their access revoked by invalidating their credentials, and thus loosing access to their funds.
```
```suggestion
The `VaultWithdraw` transaction does not respect the Permissioned Domain rules. In other words, any account that holds the shares of the vault can withdraw them. This is to avoid a situation where a depositor deposits assets to a private vault to then have their access revoked by invalidating their credentials, and thus losing access to their funds.
```

oeggert commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-withdraw.md:
```diff
@@ -0,0 +1,70 @@
+
+There are two ways to specify the transaction `Amount` field:
+
+| Withdraw | Redeem |
```
```suggestion
| Specify Assets | Specify Shares |
```
This is maybe nitpicky, but redeeming feels like the more accurate way to explain what's happening with the shares and assets. You're never withdrawing without redeeming.

oeggert reviewed (changes requested):
Made some suggestions. It might seem like a lot, but a good chunk are just to keep the language for things consistent.

maria-robobug commented:
@oeggert addressed your review comments, appreciate it! Also I've added an additional concept page for pseudo-accounts that will need reviewing. I missed this initially (assumed it was a separate amendment but it's not).

oeggert commented on docs/xls-65d-single-asset-vault/concepts/pseudo-account.md:
```diff
@@ -0,0 +1,26 @@
+
+- **Lending Protocol**: While still in development, this protocol is expected to use pseudo-accounts to hold loan funds.
+
+A pseudo-account has strict limitations. It cannot receive payments from other accounts, cannot send transactions, as it has no signing authority, and exists solely to store or issue assets, such as vault shares.
```
```suggestion
A pseudo-account has strict limitations. It cannot receive payments from other accounts, cannot send transactions since it has no signing authority, and exists solely to store or issue assets.
```
Nitpick. Just felt like a lot of comma pauses.

oeggert reviewed (approved):
Only had a couple comments, but otherwise looks good--approving.

maria-robobug commented on docs/xls-65d-single-asset-vault/reference/transactions/vault-create.md:
```diff
@@ -0,0 +1,94 @@
+  "LastLedgerSequence": 7108682,
+  "Sequence": 8,
+  "Data": "5468697320697320617262697472617279206D657461646174612061626F757420746865207661756C742E",
+  "Asset": {
+    "currency": "USD",
+    "issuer": "rIssuer1234567890abcdef1234567890abcdef",
+    "value": "1000"
```
I grabbed that from the JSON example in the spec, but doesn't seem right. Yeah you're right, I'll remove "value" from the JSON.
